### PR TITLE
feat: shadow colors

### LINKED ISSUE #1118: New text shadow support
Add the support of the new `shadow_color` component property from 24w44a in [Style](https://javadoc.io/doc/net.kyori/adventure-api/latest/net/kyori/adventure/text/format/Style.html) and MiniMessage's [standard tags](https://javadoc.io/doc/net.kyori/adventure-text-minimessage/latest/net/kyori/adventure/text/minimessage/tag/standard/StandardTags.html)

### LINKED ISSUE #1119: Add support for shadow_color to Style

### LINKED ISSUE #1120: Add support for shadow_color to MiniMessage

### PR DESCRIPTION
**Pull request by kashike**

### CODE REVIEW

File: api/src/main/java/net/kyori/adventure/text/format/ShadowColor.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * SOFTWARE.
+ */
+package net.kyori.adventure.text.format;
+
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+
+public interface ShadowColor extends StyleBuilderApplicable {
+  @Contract(pure = true)
+  static ShadowColor shadowColor(final int value) {
```
Comment (MiniDigger) [inline api/src/main/java/net/kyori/adventure/text/format/ShadowColor.java]: should document the format of that int
and maybe we can have overloads that take a textcolor + alpha?

File: api/src/main/java/net/kyori/adventure/text/format/ShadowColor.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    return new ShadowColorImpl(value);
+  }
+
+  int value();
+
+  @Override
+  default void styleApply(final Style.@NotNull Builder style) {
+    style.color(this);
```
Comment (MiniDigger) [inline api/src/main/java/net/kyori/adventure/text/format/ShadowColor.java]:
```suggestion
    style.shadowColor(this);
```

**Comment by kezz** [comment]
We need a concept of "no shadow" that exists outside of creating a shadow colour with a magic number value. I'm also not sure I like the ShadowColor object existing as it does, I think I'd prefer it to be an interface implemented by either a singleton "none" object or a textcolor + alpha value. As we need some sort of ARGBLike interface (most platforms already have a colour + alpha object), I think we should try and ensure we don't duplicate API here.

Ideally I think the API would allow for all of the following:
```java
style.shadowColor(ShadowColor.NONE);
style.shadowColor(ShadowColor.shadowColor(NamedTextColor.RED, 0.4);
style.shadowColor(NamedTextColor.RED);
style.shadowColor(myBukkitColorObject);
style.shadowColor(ShadowColor.shadowColor(0xRRGGBBAA);
```

File: text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/tag/standard/ShadowColorTagResolver.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  ShadowColorTagResolver() {
+  }
+
+  @Override
+  public @Nullable Tag resolve(final @NotNull String name, final @NotNull ArgumentQueue args, final @NotNull Context ctx) throws ParsingException {
+    if (!this.has(name)) {
+      return null;
+    }
+
+    final String colorString = args.popOr("Expected to find a color parameter: #RRGGBBAA").lowerValue();
```
Comment (zml2008) [inline text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/tag/standard/ShadowColorTagResolver.java]: Should we support something like `<shadow:red:0.5>` for this?

**Comment by zml2008** [comment]
this will need serializer support at some point too

**Comment by zml2008** [comment]
another thing we should add: a `<!shadow>` MM tag to disable shadows

**Comment by zml2008** [comment]
I think this has everything on my list now

File: api/src/main/java/net/kyori/adventure/text/Component.java [hunk 4]
```diff
@@ -2216,6 +2223,33 @@ default void detectCycle(final @NotNull Component that) {
...
+  /**
+   * Sets the shadow color of this component.
+   *
+   * @param argb the color
+   * @return a component
+   * @since 4.18.0
+   */
+  @Contract(pure = true)
+  @Override
+  default @NotNull Component shadowColor(final @Nullable ARGBLike argb) {
```
Comment (kezz) [inline api/src/main/java/net/kyori/adventure/text/Component.java:2235]: Is it worth mentioning here, or in the docs, that setting this to null doesn't _remove_ the shadow colour, it just restores it to default?

File: text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/tag/standard/StandardTags.java [hunk 5]
```diff
@@ -275,6 +276,18 @@ public static TagResolver transition() {
     return PrideTag.RESOLVER;
   }
 
+  /**
+   * Get a resolver for the {@value ShadowColorTag#SHADOW_COLOR} tags.
+   *
+   * <p>This tag support both hex string</p>
+   *
+   * @return a resolver for the {@value ShadowColorTag#SHADOW_COLOR} tags
+   * @since 4.10.0
```
Comment (kezz) [inline text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/tag/standard/StandardTags.java]:
```suggestion
   * @since 4.18.0
```

**Review by kezz — DISMISSED** [review]
Minor changes, otherwise this looks fab

**Review by zml2008 — APPROVED** [review]

**Pull request merged**